src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/DataSource.scala

```diff
@@ -112,43 +113,103 @@ class DataSource(graphFactory: OrientGraphFactory)(implicit executionContext: Ex
   def shutdown() = graphFactory.close()
 }
 
-class RawlsTransaction(val graph: OrientGraph, dataSource: DataSource) {
-  private val rollbackOnly = new AtomicBoolean(false)
-
+object RawlsTransaction {
   /**
    * We have to use JVM locks because Orient refuses to lock remote databases
    */
-  var elemLocks = TrieMap[Element, ReentrantReadWriteLock]()
+  var wsLocks = TrieMap[WorkspaceName, StampedLock]()
```

Comment:
This has been moved out of the class (a grossly stupid bug!). The lock type has also been changed to `StampedLock` as `ReentrantReadWriteLock` is... reentrant, thus tracking locks per-thread. I'm assuming that txns can span multiple threads, so per-thread locking is bad.